pkg/kubelet/config/file_test.go

```diff
@@ -125,72 +210,207 @@ func TestReadPodsFromFile(t *testing.T) {
+		func() {
+			dirName, err := utiltesting.MkTmpdir("dir-test")
+			fileName := fmt.Sprintf("%s_%d", fileNamePre, index)
+			if err != nil {
+				t.Fatalf("Unable to create temp dir: %v", err)
+			}
+			defer os.RemoveAll(dirName)
+
+			var file *os.File
+			lock := &sync.Mutex{}
+			ch := make(chan interface{})
+			finishEdit := make(chan struct{})
+			func() {
+				lock.Lock()
+				defer lock.Unlock()
+				file = testCase.writeToFile(dirName, fileName, t)
+			}()
+
+			if watchDir {
+				NewSourceFile(dirName, hostname, time.Second, ch)
+				defer func() {
+					<-finishEdit
+					// Remove the file
+					deleteFile(dirName, fileName, ch, t)
+				}()
+			} else {
+				NewSourceFile(file.Name(), hostname, time.Second, ch)
+			}
+			testCase.expected.Pods[0].Spec.Containers[0].Image = "test/newImage"
+
+			changeFile := func() {
+				// Edit the file content
+				lock.Lock()
+				defer lock.Unlock()
+				pod := testCase.pod.(*api.Pod)
+				pod.Spec.Containers[0].Image = "test/newImage"
+				testCase.pod = pod
+				testCase.writeToFile(dirName, fileName, t)
+
+				// Change the file name
+				if watchDir {
+					from := fileName
+					fileName = fileName + "_ch"
+					if err = changeFileName(dirName, from, fileName); err != nil {
+						t.Errorf("Fail to change file name: %s", err)
+					}
+				}
+				finishEdit <- struct{}{}
+			}
+			go changeFile()
+
+			timer := time.After(5 * time.Second)
+			for {
+				select {
+				case got := <-ch:
+					update := got.(kubetypes.PodUpdate)
+					for _, pod := range update.Pods {
+						if errs := validation.ValidatePod(pod); len(errs) > 0 {
+							t.Errorf("%s: Invalid pod %#v, %#v", testCase.desc, pod, errs)
+						}
+					}
+					if api.Semantic.DeepEqual(testCase.expected, update) {
+						return
+					}
+				case <-timer:
+					t.Fatalf("%s: Expected update, timeout instead", testCase.desc)
+				}
+			}
+		}()
 	}
-	expectEmptyChannel(t, ch)
 }
 
-func TestExtractFromEmptyDir(t *testing.T) {
-	dirName, err := utiltesting.MkTmpdir("file-test")
-	if err != nil {
-		t.Fatalf("Unexpected error: %v", err)
-	}
-	defer os.RemoveAll(dirName)
+func deleteFile(dir, file string, ch chan interface{}, t *testing.T) {
+	finishDel := make(chan struct{})
+	defer func() {
+		<-finishDel
+	}()
 
-	ch := make(chan interface{}, 1)
-	c := sourceFile{dirName, "localhost", ch}
-	err = c.extractFromPath()
-	if err != nil {
-		t.Fatalf("Unexpected error: %v", err)
-	}
+	go func() {
+		path := filepath.Join(dir, file)
+		err := os.Remove(path)
+		finishDel <- struct{}{}
+		if err != nil {
+			t.Errorf("Unable to remove test file %s: %s", path, err)
+		}
+	}()
 
-	update := (<-ch).(kubetypes.PodUpdate)
-	expected := CreatePodUpdate(kubetypes.SET, kubetypes.FileSource)
-	if !api.Semantic.DeepEqual(expected, update) {
-		t.Errorf("Expected %#v, Got %#v", expected, update)
+	timer := time.After(5 * time.Second)
+	for {
+		select {
+		case got := <-ch:
+			update := got.(kubetypes.PodUpdate)
+			if len(update.Pods) == 0 {
```

Comment:
Thanks for your advice. I fixed the issues you mentioned and refactored some test code to make it more readable. Please take a look, thanks.